src/java/org/apache/cassandra/tools/NodeTool.java

```diff
@@ -333,206 +211,52 @@ protected void err(Throwable e)
-        private String username = EMPTY;
-
-        @Option(type = OptionType.GLOBAL, name = {"-pw", "--password"}, description = "Remote jmx agent password")
-        private String password = EMPTY;
-
-        @Option(type = OptionType.GLOBAL, name = {"-pwf", "--password-file"}, description = "Path to the JMX password file")
-        private String passwordFilePath = EMPTY;
-
-        @Option(type = OptionType.GLOBAL, name = { "-pp", "--print-port"}, description = "Operate in 4.0 mode with hosts disambiguated by port number", arity = 0)
-        protected boolean printPort = false;
-
-        private INodeProbeFactory nodeProbeFactory;
-        protected Output output;
-
-        @Override
-        public void run(INodeProbeFactory nodeProbeFactory, Output output)
+        public CassandraCliFactory(INodeProbeFactory nodeProbeFactory, Output output)
         {
+            this.fallback = CommandLine.defaultFactory();
             this.nodeProbeFactory = nodeProbeFactory;
             this.output = output;
-            runInternal();
         }
 
-        public void runInternal()
+        public <K> K create(Class<K> cls)
         {
-            if (isNotEmpty(username)) {
-                if (isNotEmpty(passwordFilePath))
-                    password = readUserPasswordFromFile(username, passwordFilePath);
-
-                if (isEmpty(password))
-                    password = promptAndReadPassword();
-            }
-
-            try (NodeProbe probe = connect())
-            {
-                execute(probe);
-                if (probe.isFailed())
-                    throw new RuntimeException("nodetool failed, check server logs");
-            }
-            catch (IOException e)
-            {
-                throw new RuntimeException("Error while closing JMX connection", e);
-            }
-
-        }
-
-        private String readUserPasswordFromFile(String username, String passwordFilePath) {
-            String password = EMPTY;
-
-            File passwordFile = new File(passwordFilePath);
-            try (Scanner scanner = new Scanner(passwordFile.toJavaIOFile()).useDelimiter("\\s+"))
+            try
             {
-                while (scanner.hasNextLine())
+                K bean = this.fallback.create(cls);
+                Class<?> beanClass = bean.getClass();
+                do
                 {
-                    if (scanner.hasNext())
+                    Field[] fields = beanClass.getDeclaredFields();
+                    for (Field field : fields)
                     {
-                        String jmxRole = scanner.next();
-                        if (jmxRole.equals(username) && scanner.hasNext())
-                        {
-                            password = scanner.next();
-                            break;
-                        }
+                        if (!field.isAnnotationPresent(Inject.class))
+                            continue;
+                        field.setAccessible(true);
```

Comment:
Can we somehow make sure there is only one field ever for each Output and node probe factory? Is there any reason why a particular class would have more than one field of each?

Reply:
We can add a test for this for all the commands we have, doing the checks in a production path might be difficult to track as all such exceptions are runtime exceptions and I'm not sure about the code coverage in the future.

Reply:
yeah nevermind ...